wiki/PathPlanning.md

```diff
@@ -0,0 +1,67 @@
+
+### Current Implementation
+1. Remove the first point in the path if it has been reached (distance < some threshold)
+2. The velocity is computed by filtering the difference between the current position and the target position through a decoupled PID controller.
+3. The parameters of the filter are taken directly from the interface, depending on the PID type needed (e.g. the keeper will have different PID values than default).
+
+## Collision Detector
+Contains helper functions to determine:
+- robot collisions (and their position - currently returns only the collision with the first robot ID, consider taking the closest collision)
+- collision with the defense area (and its position)
+- whether a point is in field or not
+
+These helper functions are just a wrapper for the field computation functions.
+
+## Integration
+The main module just creates all the above modules and is responsible for their life cycle + calling them in the right order. Also, it incorporates some extra checks:
+1. If the target is outside of the field, no path is computed.
+2. If the robot is close to the target, but can't reach it due to collisions, it will not ram the obstacle (no further path)
+3. A path is recomputed if either one of these conditions is true:
+* there is no path
+* the target changed
+* a collision is detected between the moving robot and its next path way point
+
+## Improvements
+Consider experimenting with / researching different algorithms for different tasks. Consider also using different algorithms for each robot (and think before whether it would be better to have separate algorithms for each robot or not). Ideas:
+1. Path Planning
+- Voronoi (slow for many robots, also large avoidance - an implementation already exists in the code base)
+- Split the search area into grids and check grid occupancy
+- RRT* augmented with Particle Swarm (or other RTT implementations - many teams use RTT for their path planning)
+- Magnetic field lines (basically smart picking for RTT - draw magnetic field lines from the current position to the target and sample points from these lines to try to build a path)
+2. Path Tracking
+- Pure Pursuit (use a look ahead distance to try to track a path)
+- Fast Marching (personally I didn't understand this, but there are papers on it)
+- Bezier (interpolate the path points and take the path derivative(?) to track it)
+- Potential Field (each obstacle gives a repelling force proportional to distance, while the goal has an attracting force - maybe combine with planning)
+- Couple the PID controllers together - also taking into account the robots maximum velocity
+- Bang bang control (accelerate to maximum velocity and slow down when needed - Rolf may want to write some more details here)
```

Comment:
@rolfvdhulst

Reply:
This is actually under the wrong header kind of since it's more a path generation technique than a path tracking.

Reply:
Don't you already need a path to do this? Doing collision checking and everything else

Reply:
No, if you are under that impression you're completely misunderstanding the premise of bangbang trajectories. Multiple other teams are using them as a concept to generate smooth trajectories quickly, rather than performing control.

Reply:
Actual bang-bang control of our motors is a very bad idea, and I can probably write you an essay on why that is the case.